In [83]:
# Импортируем библиотеки - это как подключение инструментов для работы
import pandas as pd  # Для работы с таблицами данных
from sklearn.model_selection import train_test_split  # Функция для разделения данных

In [84]:
# pd.read_csv() - функция читает CSV-файл и создает DataFrame. Возвращает табличную структуру данных.
df = pd.read_csv('airline_passenger_satisfaction.csv')
# pd.set_option() - функция устанавливает параметры отображения pandas. 'display.max_columns', None означает показывать все столбцы без ограничений.
pd.set_option('display.max_columns', None)
# head() - метод возвращает первые 5 строк DataFrame (по умолчанию), полезен для быстрого просмотра структуры данных.
df.head()

,Unnamed: 0,id,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,Food and drink,Online boarding,Seat comfort,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,satisfaction
0,0,70172,Male,Loyal Customer,13,Personal Travel,Eco Plus,460,3,4,3,1,5,3,5,5,4,3,4,4,5,5,25,18.0,neutral or dissatisfied
1,1,5047,Male,disloyal Customer,25,Business travel,Business,235,3,2,3,3,1,3,1,1,1,5,3,1,4,1,1,6.0,neutral or dissatisfied
2,2,110028,Female,Loyal Customer,26,Business travel,Business,1142,2,2,2,2,5,5,5,5,4,3,4,4,4,5,0,0.0,satisfied
3,3,24026,Female,Loyal Customer,25,Business travel,Business,562,2,5,5,5,2,2,2,2,2,5,3,1,4,2,11,9.0,neutral or dissatisfied
4,4,119299,Male,Loyal Customer,61,Business travel,Business,214,3,3,3,3,4,5,5,3,3,4,4,3,3,3,0,0.0,satisfied


In [85]:
# drop() - метод удаляет указанные столбцы или строки из DataFrame

# columns - список столбцов для удаления
# inplace=True - изменяет исходный DataFrame без создания копии

df.drop(columns = ['id', 'Unnamed: 0'], inplace=True)
df.dropna(inplace=True)
df.head()

,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,Food and drink,Online boarding,Seat comfort,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,satisfaction
0,Male,Loyal Customer,13,Personal Travel,Eco Plus,460,3,4,3,1,5,3,5,5,4,3,4,4,5,5,25,18.0,neutral or dissatisfied
1,Male,disloyal Customer,25,Business travel,Business,235,3,2,3,3,1,3,1,1,1,5,3,1,4,1,1,6.0,neutral or dissatisfied
2,Female,Loyal Customer,26,Business travel,Business,1142,2,2,2,2,5,5,5,5,4,3,4,4,4,5,0,0.0,satisfied
3,Female,Loyal Customer,25,Business travel,Business,562,2,5,5,5,2,2,2,2,2,5,3,1,4,2,11,9.0,neutral or dissatisfied
4,Male,Loyal Customer,61,Business travel,Business,214,3,3,3,3,4,5,5,3,3,4,4,3,3,3,0,0.0,satisfied


In [86]:
# df.drop(columns='satisfaction') - удаляет столбец 'satisfaction', создавая матрицу признаков X

X = df.drop("satisfaction", axis=1)
X.head()

,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,Food and drink,Online boarding,Seat comfort,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes
0,Male,Loyal Customer,13,Personal Travel,Eco Plus,460,3,4,3,1,5,3,5,5,4,3,4,4,5,5,25,18.0
1,Male,disloyal Customer,25,Business travel,Business,235,3,2,3,3,1,3,1,1,1,5,3,1,4,1,1,6.0
2,Female,Loyal Customer,26,Business travel,Business,1142,2,2,2,2,5,5,5,5,4,3,4,4,4,5,0,0.0
3,Female,Loyal Customer,25,Business travel,Business,562,2,5,5,5,2,2,2,2,2,5,3,1,4,2,11,9.0
4,Male,Loyal Customer,61,Business travel,Business,214,3,3,3,3,4,5,5,3,3,4,4,3,3,3,0,0.0


In [87]:
y = df['satisfaction'] # Извлекает целевую переменную (то, что мы хотим предсказать)
y.head()

0    neutral or dissatisfied
1    neutral or dissatisfied
2                  satisfied
3    neutral or dissatisfied
4                  satisfied
Name: satisfaction, dtype: str

In [88]:
# Преобразуем текстовые признаки в числовые
X = pd.get_dummies(X, drop_first=True)

In [89]:
# ПРАВИЛЬНОЕ разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(
    X, y,                    # Наши данные
    test_size=0.2,          # 20% данных для тестирования
    random_state=42,        # Фиксируем случайность для воспроизводимости
    stratify=y              # Сохраняем пропорции классов в обеих выборках
)

### Random Forest (Домашнее задание)

In [90]:
from sklearn.ensemble import RandomForestClassifier # RandomForestClassifier - алгоритм "случайный лес", который объединяет множество деревьев решений
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, fbeta_score, confusion_matrix  # Импортируем метрики точности

model_random_forest = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)
print("Начинаем обучение модели RandomForestClassifier")

# Обучаем модель
model_random_forest.fit(X_train, y_train)

print("Модель RandomForestClassifier обучена")

Начинаем обучение модели RandomForestClassifier
Модель RandomForestClassifier обучена


In [91]:
# Делаем предсказания
predictions_rf = model_random_forest.predict(X_test)

In [92]:
# Вычисление всех метрик качества модели
accuracy = accuracy_score(y_test, predictions_rf)  # accuracy_score() - сравнивает истинные значения y_test с предсказаниями и возвращает долю совпадений
# pos_label='satisfied' → считаем метрику только для одного класса. Только для бинарной классификации (satisfied vs neutral or dissatisfied).
# В таком режиме метрика отвечает на вопрос: «Насколько хорошо модель умеет находить именно класс 'satisfied'?»
precision = precision_score(y_test, predictions_rf, pos_label='satisfied')  # precision_score() - вычисляет точность (из предсказанных положительных, сколько действительно положительных)
recall = recall_score(y_test, predictions_rf, pos_label='satisfied')  # recall_score() - вычисляет полноту (из всех положительных случаев, сколько мы нашли)
f1 = f1_score(y_test, predictions_rf, pos_label='satisfied')  # f1_score() - гармоническое среднее между precision и recall

# average='weighted' → берём оба класса и усредняем их результат, учитывая их количество.
# average='weighted' отвечает на вопрос: «В среднем по всем классам, насколько хорошо работает модель»
f_beta_05 = fbeta_score(y_test, predictions_rf, beta=0.5, average='weighted')  # fbeta_score() с beta=0.5 - больше веса precision
f_beta_2 = fbeta_score(y_test, predictions_rf, beta=2.0, average='weighted')  # fbeta_score() с beta=2.0 - больше веса recall

print("📊 РЕЗУЛЬТАТЫ ОЦЕНКИ МОДЕЛИ RandomForestClassifier:")
print("="*50)
print(f"Accuracy (Точность):           {accuracy:.4f} ({accuracy*100:.2f}%)")
print(f"Precision (Точность полож.):   {precision:.4f} ({precision*100:.2f}%)")
print(f"Recall (Полнота):              {recall:.4f} ({recall*100:.2f}%)")
print(f"F1-Score:                      {f1:.4f} ({f1*100:.2f}%)")
print(f"F-Beta (β=0.5, упор на Precis): {f_beta_05:.4f} ({f_beta_05*100:.2f}%)")
print(f"F-Beta (β=2.0, упор на Recall): {f_beta_2:.4f} ({f_beta_2*100:.2f}%)")

print(f"\n📝 Интерпретация результатов:")
print(f"• Из {len(y_test)} пассажиров модель правильно классифицировала {int(accuracy * len(y_test))}")
print(f"• Общая точность составляет {accuracy*100:.2f}%")

📊 РЕЗУЛЬТАТЫ ОЦЕНКИ МОДЕЛИ RandomForestClassifier:
Accuracy (Точность):           0.9621 (96.21%)
Precision (Точность полож.):   0.9706 (97.06%)
Recall (Полнота):              0.9410 (94.10%)
F1-Score:                      0.9556 (95.56%)
F-Beta (β=0.5, упор на Precis): 0.9621 (96.21%)
F-Beta (β=2.0, упор на Recall): 0.9620 (96.20%)

📝 Интерпретация результатов:
• Из 20719 пассажиров модель правильно классифицировала 19933
• Общая точность составляет 96.21%


### KNeighborsClassifier (Домашнее задание)

In [93]:
from sklearn.neighbors import KNeighborsClassifier # KNeighborsClassifier - алгоритм k ближайших соседей n_neighbors=5 - количество соседей для классификации
from sklearn.preprocessing import StandardScaler

model_kn_neighbors_classifier = KNeighborsClassifier(
    n_neighbors=5
)
print("Начинаем обучение модели KNeighborsClassifier")

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Обучаем модель
model_kn_neighbors_classifier.fit(X_train_scaled, y_train)
print("Модель KNeighborsClassifier обучена")

Начинаем обучение модели KNeighborsClassifier
Модель KNeighborsClassifier обучена


In [94]:
# Делаем предсказания
predictions_knn = model_kn_neighbors_classifier.predict(X_test_scaled)

In [95]:
# Вычисление всех метрик качества модели
accuracy = accuracy_score(y_test, predictions_knn)  # accuracy_score() - сравнивает истинные значения y_test с предсказаниями и возвращает долю совпадений
# pos_label='satisfied' → считаем метрику только для одного класса. Только для бинарной классификации (satisfied vs neutral or dissatisfied).
# В таком режиме метрика отвечает на вопрос: «Насколько хорошо модель умеет находить именно класс 'satisfied'?»
precision = precision_score(y_test, predictions_knn, pos_label='satisfied')  # precision_score() - вычисляет точность (из предсказанных положительных, сколько действительно положительных)
recall = recall_score(y_test, predictions_knn, pos_label='satisfied')  # recall_score() - вычисляет полноту (из всех положительных случаев, сколько мы нашли)
f1 = f1_score(y_test, predictions_knn, pos_label='satisfied')  # f1_score() - гармоническое среднее между precision и recall

# average='weighted' → берём оба класса и усредняем их результат, учитывая их количество.
# average='weighted' отвечает на вопрос: «В среднем по всем классам, насколько хорошо работает модель»
f_beta_05 = fbeta_score(y_test, predictions_knn, beta=0.5, average='weighted')  # fbeta_score() с beta=0.5 - больше веса precision
f_beta_2 = fbeta_score(y_test, predictions_knn, beta=2.0, average='weighted')  # fbeta_score() с beta=2.0 - больше веса recall

print("📊 РЕЗУЛЬТАТЫ ОЦЕНКИ МОДЕЛИ KNeighborsClassifier:")
print("="*50)
print(f"Accuracy (Точность):           {accuracy:.4f} ({accuracy*100:.2f}%)")
print(f"Precision (Точность полож.):   {precision:.4f} ({precision*100:.2f}%)")
print(f"Recall (Полнота):              {recall:.4f} ({recall*100:.2f}%)")
print(f"F1-Score:                      {f1:.4f} ({f1*100:.2f}%)")
print(f"F-Beta (β=0.5, упор на Precis): {f_beta_05:.4f} ({f_beta_05*100:.2f}%)")
print(f"F-Beta (β=2.0, упор на Recall): {f_beta_2:.4f} ({f_beta_2*100:.2f}%)")

print(f"\n📝 Интерпретация результатов:")
print(f"• Из {len(y_test)} пассажиров модель правильно классифицировала {int(accuracy * len(y_test))}")
print(f"• Общая точность составляет {accuracy*100:.2f}%")

📊 РЕЗУЛЬТАТЫ ОЦЕНКИ МОДЕЛИ KNeighborsClassifier:
Accuracy (Точность):           0.9257 (92.57%)
Precision (Точность полож.):   0.9481 (94.81%)
Recall (Полнота):              0.8766 (87.66%)
F1-Score:                      0.9109 (91.09%)
F-Beta (β=0.5, упор на Precis): 0.9260 (92.60%)
F-Beta (β=2.0, упор на Recall): 0.9253 (92.53%)

📝 Интерпретация результатов:
• Из 20719 пассажиров модель правильно классифицировала 19180
• Общая точность составляет 92.57%


### Logistic Regression (Домашнее задание)

In [96]:
from sklearn.linear_model import LogisticRegression # LogisticRegression - алгоритм логистической регрессии max_iter=10000 - максимальное количество итераций для сходимости

model_logistic_regression = LogisticRegression(
    max_iter=10000
)
print("Начинаем обучение модели LogisticRegression")

# Обучаем модель
model_logistic_regression.fit(X_train_scaled, y_train)

print("Модель LogisticRegression обучена")

Начинаем обучение модели LogisticRegression
Модель LogisticRegression обучена


In [97]:
# Делаем предсказания
predictions_lr = model_logistic_regression.predict(X_test_scaled)

In [98]:
# Вычисление всех метрик качества модели
accuracy = accuracy_score(y_test, predictions_lr)  # accuracy_score() - сравнивает истинные значения y_test с предсказаниями и возвращает долю совпадений
# pos_label='satisfied' → считаем метрику только для одного класса. Только для бинарной классификации (satisfied vs neutral or dissatisfied).
# В таком режиме метрика отвечает на вопрос: «Насколько хорошо модель умеет находить именно класс 'satisfied'?»
precision = precision_score(y_test, predictions_lr, pos_label='satisfied')  # precision_score() - вычисляет точность (из предсказанных положительных, сколько действительно положительных)
recall = recall_score(y_test, predictions_lr, pos_label='satisfied')  # recall_score() - вычисляет полноту (из всех положительных случаев, сколько мы нашли)
f1 = f1_score(y_test, predictions_lr, pos_label='satisfied')  # f1_score() - гармоническое среднее между precision и recall

# average='weighted' → берём оба класса и усредняем их результат, учитывая их количество.
# average='weighted' отвечает на вопрос: «В среднем по всем классам, насколько хорошо работает модель»
f_beta_05 = fbeta_score(y_test, predictions_lr, beta=0.5, average='weighted')  # fbeta_score() с beta=0.5 - больше веса precision
f_beta_2 = fbeta_score(y_test, predictions_lr, beta=2.0, average='weighted')  # fbeta_score() с beta=2.0 - больше веса recall

print("📊 РЕЗУЛЬТАТЫ ОЦЕНКИ МОДЕЛИ LogisticRegression:")
print("="*50)
print(f"Accuracy (Точность):           {accuracy:.4f} ({accuracy*100:.2f}%)")
print(f"Precision (Точность полож.):   {precision:.4f} ({precision*100:.2f}%)")
print(f"Recall (Полнота):              {recall:.4f} ({recall*100:.2f}%)")
print(f"F1-Score:                      {f1:.4f} ({f1*100:.2f}%)")
print(f"F-Beta (β=0.5, упор на Precis): {f_beta_05:.4f} ({f_beta_05*100:.2f}%)")
print(f"F-Beta (β=2.0, упор на Recall): {f_beta_2:.4f} ({f_beta_2*100:.2f}%)")

print(f"\n📝 Интерпретация результатов:")
print(f"• Из {len(y_test)} пассажиров модель правильно классифицировала {int(accuracy * len(y_test))}")
print(f"• Общая точность составляет {accuracy*100:.2f}%")

📊 РЕЗУЛЬТАТЫ ОЦЕНКИ МОДЕЛИ LogisticRegression:
Accuracy (Точность):           0.8717 (87.17%)
Precision (Точность полож.):   0.8664 (86.64%)
Recall (Полнота):              0.8323 (83.23%)
F1-Score:                      0.8490 (84.90%)
F-Beta (β=0.5, упор на Precis): 0.8714 (87.14%)
F-Beta (β=2.0, упор на Recall): 0.8715 (87.15%)

📝 Интерпретация результатов:
• Из 20719 пассажиров модель правильно классифицировала 18061
• Общая точность составляет 87.17%


## Сравнение результатов моделей

In [99]:
results = pd.DataFrame({
    "Model": [
        "RandomForestClassifier",
        "KNeighborsClassifier",
        "LogisticRegression"
    ],
    "Accuracy": [
        accuracy_score(y_test, predictions_rf),
        accuracy_score(y_test, predictions_knn),
        accuracy_score(y_test, predictions_lr)
    ],
    "Precision": [
        precision_score(y_test, predictions_rf, pos_label="satisfied"),
        precision_score(y_test, predictions_knn, pos_label="satisfied"),
        precision_score(y_test, predictions_lr, pos_label="satisfied")
    ],
    "Recall": [
        recall_score(y_test, predictions_rf, pos_label="satisfied"),
        recall_score(y_test, predictions_knn, pos_label="satisfied"),
        recall_score(y_test, predictions_lr, pos_label="satisfied")
    ],
    "F1-score": [
        f1_score(y_test, predictions_rf, pos_label="satisfied"),
        f1_score(y_test, predictions_knn, pos_label="satisfied"),
        f1_score(y_test, predictions_lr, pos_label="satisfied")
    ]
})

results.sort_values(by="Accuracy", ascending=False)

,Model,Accuracy,Precision,Recall,F1-score
0,RandomForestClassifier,0.962064,0.970592,0.940973,0.955553
1,KNeighborsClassifier,0.925720,0.948085,0.876601,0.910943
2,LogisticRegression,0.871712,0.866435,0.832275,0.849012


В рамках домашнего задания были обучены и оценены три модели:
RandomForestClassifier, KNeighborsClassifier и LogisticRegression.

По результатам оценки:

- RandomForestClassifier показал лучший результат: Accuracy = 96.21%.
- KNeighborsClassifier показал результат ниже: Accuracy = 92.57%.
- LogisticRegression показала самый низкий результат среди трех моделей: Accuracy = 87.17%.

Лучшей моделью для данной задачи является RandomForestClassifier. Вероятно, это связано с тем, что Random Forest хорошо работает с табличными данными и способен находить сложные нелинейные зависимости между признаками.

KNeighborsClassifier также показал хороший результат, но он чувствителен к масштабу данных, поэтому для него было использовано масштабирование признаков.

LogisticRegression показала более слабый результат, так как это более простая линейная модель, и она может хуже описывать сложные зависимости в данных.

Итоговый вывод: для задачи предсказания удовлетворенности пассажиров авиакомпании лучше всего подошла модель RandomForestClassifier.